### PR TITLE
Add instance validation check for ContentDetectionResponseObject for new_result in process, and empty content response from detector

### PR DESCRIPTION
**Pull request by swith004**
## What this PR does
This PR adds an instance validation check to processed_results for processed chat completion results.

Currently, there isn't a [validation check if the new processed result is an error](https://github.com/foundation-model-stack/vllm-detector-adapter/blob/main/vllm_detector_adapter/generative_detectors/base.py#L424-L450), meaning that if the new result is in fact an instance of ErrorResponse, then the returned error [error may get returned](https://github.com/foundation-model-stack/vllm-detector-adapter/blob/main/vllm_detector_adapter/generative_detectors/base.py#L450), gets added to the processed_results list with the Error in the list, causing a pydantic error. 

This PR addresses that by checking if the new_result instance is an instance of an ErrorResponse, and returns that error response. Otherwise, the new_result is added to the processed_results list. 
Closes [1773](https://github.ibm.com/ai-foundation/watson-fm-stack-tracker/issues/1773)

This PR also adds a check on if the content response is empty, and if so, this should also produce an ErrorResponse.
Closes [70](https://github.com/foundation-model-stack/vllm-detector-adapter/issues/70)

## How this was tested
- This PR includes a unit test that validates that the ErrorResponse is propagated for issue 1773
- A unit test is also included to validate an empty content response produces an ErrorResponse for 70

### CODE REVIEW

File: vllm_detector_adapter/generative_detectors/base.py [hunk 1]
```diff
@@ -447,6 +447,24 @@ async def content_analysis(
...
+                    logger.error(
+                        f"[content_analysis] new_result is not ContentsDetectionResponseObject: {repr(new_result)}"
+                    )
+
+                    return ErrorResponse(
+                        type="BadRequestError",
+                        message=f"Result at index {result_idx} is not a valid ContentsDetectionResponseObject.",
+                        code=HTTPStatus.BAD_REQUEST.value,
+                    )
+
```
[the remark this answers is not in the document]
Comment (gkumbhat) [reply vllm_detector_adapter/generative_detectors/base.py:467]: +1 ^

File: vllm_detector_adapter/generative_detectors/base.py [hunk 2]
```diff
@@ -447,6 +447,24 @@ async def content_analysis(
...
 
+                # Verify whether the new_result is the correct response instance (an error could make the type an invalid value, such as a tuple, etc)
+                if not (
+                    isinstance(new_result, list)
+                    and all(
+                        isinstance(item, (ContentsDetectionResponseObject, dict))
+                        for item in new_result
+                    )
+                ):
+                    logger.error(
```
Comment (gkumbhat) [inline vllm_detector_adapter/generative_detectors/base.py:458]: printing `new_result` with `error` can be a problem, since it can leak user data. Lets move this to `debug` level log.

File: vllm_detector_adapter/generative_detectors/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (swith004) [reply vllm_detector_adapter/generative_detectors/base.py:467]: Thanks for the review @evaline-ju  - 

1. If I understand your comment correctly, I believe you're referring to the method itself in[ protocol.py](https://github.com/foundation-model-stack/vllm-detector-adapter/blob/main/vllm_detector_adapter/protocol.py#L96-L107) that does have an ErrorResponse that wasn't considered when returning the result. I think our concern here is that the "new result" may not be a valid instance that we can process, i.e. it could be a tuple or an invalid typem where we expect a dict or ContentsDetectionResponseObject, at least per the error stack from the pydantic issue. 
The case or error message there I thought may not apply to this bug, so I tried to capture the instance issue here with a relevant message, by checking on the. 

If it's preferred to remove the check here,  or add an additional check that checks if the instance is ErrorResponse and return that as well, I will address that next push.

2. That makes sense. I'll update the message as well.

File: vllm_detector_adapter/generative_detectors/base.py [hunk 3]
```diff
@@ -447,21 +447,25 @@ async def content_analysis(
...
                 ):
-                    logger.error(
+                    logger.debug(
                         f"[content_analysis] new_result is not ContentsDetectionResponseObject: {repr(new_result)}"
                     )
 
                     return ErrorResponse(
                         type="BadRequestError",
-                        message=f"Result at index {result_idx} is not a valid ContentsDetectionResponseObject.",
+                        message=f"Invalid result at index {result_idx}. Consider updating input and/or parameters for detections.",
```
Comment (gkumbhat) [inline vllm_detector_adapter/generative_detectors/base.py:468]: Since now we have the `ErrorResponse` handled above, what would be the case where this `if` condition we expect to reach ?

Comment (swith004) [reply vllm_detector_adapter/generative_detectors/base.py:468]: Good question. Although the errorresponse return is handled from that function, that function still doesn't handle an error that may produce a tuple or other error type slipping through the response it seems, as I determined when running the unit test I added for this error; I would get the pydantic error message that is seen in SW's error logs during my unit test. 

Therefore this if condition is the result of whatever edge case appears in the response, and will be handled in this block.

Comment (swith004) [reply vllm_detector_adapter/generative_detectors/base.py:468]: Please allow me to look for a way to move this in the [protocol](https://github.com/foundation-model-stack/vllm-detector-adapter/blob/main/vllm_detector_adapter/protocol.py#L54) file and test with the unit test with the tuple example

File: vllm_detector_adapter/protocol.py [hunk 4]
```diff
@@ -105,6 +106,18 @@ def from_chat_completion_response(
...
+            and all(
+                isinstance(item, (ContentsDetectionResponseObject, dict))
+                for item in detection_responses
+            )
+        ):
+            return ErrorResponse(
+                type="BadRequestError",
+                message=f"Invalid result. Consider updating input and/or parameters for detections.",
+                code=HTTPStatus.BAD_REQUEST.value,
+            )
```
Comment (gkumbhat) [inline vllm_detector_adapter/protocol.py]: Is this still needed, considering line `102` will already return `ErrorResponse` ?

File: vllm_detector_adapter/protocol.py [hunk 5]
```diff
@@ -82,7 +82,7 @@ def from_chat_completion_response(
             # NOTE: for providing spans, we currently consider entire generated text as a span.
             # This is because, at the time of writing, the generative guardrail models does not
             # provide specific information about input text, which can be used to deduce spans.
-            if content and isinstance(content, str):
+            if content and isinstance(content, str) and len(content.strip()) > 0:
```
Comment (gkumbhat) [inline vllm_detector_adapter/protocol.py:85]: can `len(content.strip()) > 0:` be replaced to `content.strip()`, since empty string is falsy

File: vllm_detector_adapter/protocol.py (hunk 4 above)
Comment (swith004) [reply vllm_detector_adapter/protocol.py]: From my test, this is very much needed. Line 102 does not handle the bug from software, as it appears to me that the errorresponse in that line  is based on the chat completion having content or not (the error message indicates this as well). 

A content missing error is not the error happening in the SW bug. While we don't quite know what triggers the issue, we know that a tuple is returned instead of a list or contentdetectionresponseobject, and therefore it's not handled by the above. Removing this check will cause the pytest to fail for this reason, as it won't rturn an errorresponse if a tuple is patched into the list This is very

File: tests/test_protocol.py [hunk 6]
```diff
@@ -357,3 +358,38 @@ def test_response_from_completion_response_missing_content():
...
+    )
+    scores = [0.5]
+    detection_type = "type"
+    req_content = "dummy"
+
+    # Add a patch to return a tuple during the model_dump call instead of a valid instance (dict, ContentsDetectionResponseObject)
+    # This will cause the list to have a tuple in it that should return an ErrorResponse
+    with patch(
+        "vllm_detector_adapter.protocol.ContentsDetectionResponseObject.model_dump",
+        return_value=("this", "is", "not", "valid"),
```
[the remark this answers is not in the document]
Comment (swith004) [reply tests/test_protocol.py:386]: You are right in reality this wouldn't occur. However, since I could not replicate and therefore pinpoint at which point [detection_results](https://github.com/foundation-model-stack/vllm-detector-adapter/blob/main/vllm_detector_adapter/protocol.py#L76-L95) gets a tuple in the list, I tried to hack a way to patch it in the test just to make sure the errorresponse is returned in this test environment (to at least verify the logic). 

Definitely open to exploring another hack to achieve the same results.

File: vllm_detector_adapter/protocol.py [hunk 7]
```diff
@@ -82,7 +82,7 @@ def from_chat_completion_response(
             # NOTE: for providing spans, we currently consider entire generated text as a span.
             # This is because, at the time of writing, the generative guardrail models does not
             # provide specific information about input text, which can be used to deduce spans.
-            if content and isinstance(content, str) and len(content.strip()) > 0:
+            if content and isinstance(content, str) and content.strip():
```
[the remark this answers is not in the document]
Comment (swith004) [reply vllm_detector_adapter/protocol.py:85]: Hmm, So reviewing the [full context here](https://github.com/foundation-model-stack/vllm-detector-adapter/blob/main/vllm_detector_adapter/protocol.py#L80-L85), the first check validates that content exist at all, then it checks if it's a string, and I added the check to determine if it's an empty string, so I could also close this [bug here](https://github.com/foundation-model-stack/vllm-detector-adapter/issues/70) - pending the unit test I'm also adding to verify).

Unless I'm mistaken, wouldn't reordering mess up this logic?

File: tests/test_protocol.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gkumbhat) [reply tests/test_protocol.py:386]: What happens if it is still a dict but not like `ContentsDetectionResponseObject` ?

File: vllm_detector_adapter/protocol.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gkumbhat) [reply vllm_detector_adapter/protocol.py:85]: `content` can be removed. `None` is of `NoneType` so it will fail `isinstance(content,str)` check and empty content will be taken care off by `content.strip()`.

File: tests/generative_detectors/test_base.py [hunk 8]
```diff
@@ -204,3 +207,40 @@ def test_content_analysis_success(detection_base, completion_response):
...
+    detection_type = "risk"
+
+    # Patch the process_chat_completion_with_scores and post_process_completion_results to return results
+    with (
+        patch(
+            "vllm_detector_adapter.generative_detectors.base.ChatCompletionDetectionBase.process_chat_completion_with_scores",
+            return_value=(response, scores, detection_type, {}),
+        ),
+        patch(
+            "vllm_detector_adapter.generative_detectors.base.ChatCompletionDetectionBase.post_process_completion_results",
```
Comment (gkumbhat) [inline tests/generative_detectors/test_base.py:238]: is patching this one required ?

File: vllm_detector_adapter/generative_detectors/base.py [hunk 9]
```diff
@@ -447,6 +447,13 @@ async def content_analysis(
                     )
                 )
 
+                # Verify whether the new_result is the correct is an errorresponse, and if so, returh the errorresponse
```
Comment (gkumbhat) [inline vllm_detector_adapter/generative_detectors/base.py:450]: Can we apply same changes for llama guard as well, since that one might also have same problem

Comment (swith004) [reply vllm_detector_adapter/generative_detectors/base.py:450]: good point. Let me make this change

File: tests/generative_detectors/test_base.py (hunk 8 above)
Comment (swith004) [reply tests/generative_detectors/test_base.py:238]: I thought it was based on an earlier test I made, but let me verify this...

File: vllm_detector_adapter/protocol.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (swith004) [reply vllm_detector_adapter/protocol.py:85]: Understood. - thanks

File: tests/generative_detectors/test_base.py (hunk 8 above)
Comment (swith004) [reply tests/generative_detectors/test_base.py]: thanks for checking - reviewed other unit test to match it's patching method

**Review by gkumbhat — APPROVED** [review]
Looks good to me! Thanks Shonda for all the iterations

File: vllm_detector_adapter/generative_detectors/llama_guard.py [hunk 10]
```diff
@@ -195,6 +195,13 @@ async def content_analysis(
                     )
                 )
 
+                # Verify whether the new_result is the correct is an errorresponse, and if so, return the errorresponse
+                if isinstance(new_result, ErrorResponse):
+                    logger.debug(
+                        f"[llama_guard_content_analysis] ErrorResponse returned: {repr(new_result)}"
```
Comment (gkumbhat) [inline vllm_detector_adapter/generative_detectors/llama_guard.py:201]: since one vllm only runs 1 model, adding ``llama_guard_` can be avoided

Comment (swith004) [reply vllm_detector_adapter/generative_detectors/llama_guard.py]: right; that makes sense. I pushed another update to remove that prepended debug message"
